### ISSUE TITLE
Every once in awhile you need to run stack a few times to build all your depends because it gets the order wrong

### ISSUE THREAD

**Issue by Gurkenglas** [commercialhaskell/stack#2512]
This issue report was prompted by the following conversation:

```
<MarcelineVQ> Gurkenglas: every once in awhile you need to run stack a few times to build all your depends because it gets the order wrong, I remember this happenig with cairo specifically one time in fact. This is the yaml I use to get reflex/dom started on ghc8 project, no idea if it works for windows http://lpaste.net/177279
<Gurkenglas> That sounds like at the very least stack should retry on failure?
<MarcelineVQ> hehe, I haven't tried to reproduce it to narrow down the culprit, I was just happy that stack got a little farther each time I retried building the thing it happened with and then it worked
<Gurkenglas> I'll replace those git@ links with https:// coz i dont have that key thing setup, and remove the extra-dep:true lines so itll actually have to install those
<MarcelineVQ> Oh?​ What are you trying to accomplish?
<Gurkenglas> Getting any of this to build, and I guess seeing the Hello World text in the browser at some point
<MarcelineVQ> I'm not sure you need a key to pull from a git@ just to push to, not sure though
<Gurkenglas> nope tried without got an error that say public key in it
<Gurkenglas> *says
<Gurkenglas> http://lpaste.net/3330066059314069504 Ill retry a few times I guess
<MarcelineVQ> yes I believe that was one of the ones I had to retry for as well, perhaps it's a problem when building with git
<MarcelineVQ> Gtk2HsSetup being the one I mean
<Gurkenglas> I'll put an issue on stack's github tracker, linking to your 23:58 line?
<Gurkenglas> *:23:58
<Gurkenglas> Wait this channel isn't publically logged. Can I paste all since then?
<MarcelineVQ> I think our clocks are different, feel free to paste anything I say anywhere you like though
<MarcelineVQ> The fact that it's clearly reproducable is good though
<Gurkenglas> I explicitly left out the hour, shouldn't all clocks on the internet agree on the rest? ._.
<Gurkenglas> Second try: (Pasting for here and for the ticket: http://lpaste.net/1948311493200052224 )
<MarcelineVQ> keep at it, eventually it'll build Gtk2HsSetup and then work, at least it did for me
<Gurkenglas> I think I also had something like that at some point with some stack build process?
<Gurkenglas> Sounds like it's building in parallel, didn't notice that it had to install something before something else, but the else takes a while and so other stuff has a little time to build, and once it's built it wont build that again and have time to build something else instead, until the thing that the else needs gets built by luck
<Gurkenglas> yep cairo just built
<Gurkenglas> A workaround would be to keep installing other packages that don't need that one that failed, and retrying the one that failed afterwards?
<MarcelineVQ> That seems reasonable, just keeping a list of failed packages to try again after the rest finish
<MarcelineVQ> And then retrying, and if none of them can build it's an error, but if any build then it has a new list of failed packages to try again
<Gurkenglas> Should look less like a list and more like a front along the directed acyclic graph of packages to be built
```

My terminal session starts like http://lpaste.net/6497164652642828288 (some of the `D:\Haskell\reflex-test2>stack build 2> log.txt` were Ctrl-C'd immediately), at the and of which, before I look at the terminal again, the output explodes and gives a suffix of this logfile, which is generated on each subsequent stack build try:

Edit: The previous version of this upload cut off after a few lines? [log3.txt](https://github.com/commercialhaskell/stack/files/424831/log3.txt)

**Comment by Blaisorblade** [comment]
Let me try to understand.

- in your log, stack tries building `cairo` before `gtk2hs-buildtools`, and this fails. Restarting the build after failure works if the first build manages to build `gtk2hs-buildtools`. That is, stack should build `cairo` after `gtk2hs-buildtools`. I think it should be possible to fix that, and opened #2515 for that part.
Personally, I'd like to avoid "try again" and fix this properly, unless maybe it turns out to be insanely hard.

- at the end, `ghcjs-dom` has a repeatable compile error. Conjecture: that's a correct compile error and stack has no fault. Right?

**Comment by Gurkenglas** [comment]
The first is correct, the second I don't know: https://hackage.haskell.org/package/ghcjs-dom-0.2.4.0/docs/src/GHCJS-DOM-Types.html#IsTouch , which uses the out-of-scope TouchClass, looks like it wouldn't be defined if USE_OLD_WEBKIT was, and the same goes for https://hackage.haskell.org/package/ghcjs-dom-0.2.4.0/docs/src/GHCJS-DOM-Types.html#IsWheelEvent . Maybe stack "failed to" set the USE_OLD_WEBKIT flag?

**Comment by Blaisorblade** [comment]
That suggests you should enable the flag `old-webkit`. `USE_OLD_WEBKIT ` is set by using the flag `old-webkit`, which is by default false. For stack build, that probably means adding `--flag ghcjs-dom:old-webkit`, though flags can be set in `stack.yaml`: https://docs.haskellstack.org/en/latest/yaml_configuration/#flags.

Quoting from [ghcjs-dom's cabal file](https://hackage.haskell.org/package/ghcjs-dom-0.2.4.0/ghcjs-dom.cabal):

```cabal
flag old-webkit
    description: If you have an old version of WebKitGTK+ this will get ghcjs-dom
                 to build, leaving out functions that are new or have changed.
                 This option is mainly here to get the travis builds to work.
    default: False
```

**Comment by mgsloan** [comment]
It looks like unfortunately cairo removed the build-tools from the cabal metadata: https://github.com/gtk2hs/gtk2hs/commit/3fdb125c57e853a9ea473a54e277600be7d962bd

I think the reason that the issue isn't detected when using cabal is that it is instead relying on it being build via the custom setup dep.  Added a comment here https://github.com/commercialhaskell/stack/issues/2094#issuecomment-241174321

**Comment by mgsloan** [comment]
One option is just to downgrade to `cairo-0.13.2.0`, which shouldn't have this issue.

**Comment by Blaisorblade** [comment]
That version already uses `custom-setup`; it works against nightly-2016-08-19, but otherwise I fear you might need `cairo-0.13.1.1`.
I just failed to compile `cairo-0.13.2.0` with
```
resolver: lts-2.22
packages: []
extra-deps:
- cairo-0.13.2.0
  #- gtk2hs-buildtools-0.13.1.0
```
It installed gtk2hs-buildtools-0.13.0.4 (violating the dependency in `custom-setup` since we don't support it) and didn't find `Gtk2HsSetup` (which was introduced in `0.13.1.0`). Even if I get the right version to install (working around #595) with 

```
resolver: lts-2.22
packages:
- '.'
extra-deps:
- order-maintenance-0.1.1.0
- cairo-0.13.2.0
- gtk2hs-buildtools-0.13.1.0
- Cabal-1.24.0.0
```
I get the same failure, because Setup is only built against the local DB.

Instead, latest `nightly` resolver already includes a recent enough `gtk2hs-buildtools`, so everything works with the following `stack.yaml`:

```
resolver: nightly-2016-08-19
packages: []
extra-deps:
- cairo-0.13.2.0
```
